torch/csrc/dynamo/python_compiled_autograd.cpp

```diff
@@ -0,0 +1,479 @@
+    }
+    return it->second;
+  }
+};
+
+static PyObject* the_autograd_compiler = nullptr;
+static PyObject* set_autograd_compiler(PyObject* dummy, PyObject* args);
+
+static PyObject* clear_cache(PyObject* dummy, PyObject* args) {
+  CacheNode::root()->clear();
+  Py_RETURN_NONE;
+}
+
+static PyObject* is_cache_empty(PyObject* dummy, PyObject* args) {
+  if (CacheNode::root()->is_empty()) {
+    Py_RETURN_TRUE;
+  }
+  Py_RETURN_FALSE;
+}
+
+static PyMethodDef _methods[] = {
+    {"set_autograd_compiler", set_autograd_compiler, METH_VARARGS, NULL},
+    {"clear_cache", clear_cache, METH_NOARGS, NULL},
+    {"is_cache_empty", is_cache_empty, METH_NOARGS, NULL},
+    {NULL, NULL, 0, NULL}};
+
+static struct PyModuleDef _module = {
+    PyModuleDef_HEAD_INIT,
+    "torch._C._dynamo.autograd_compiler",
+    "Hooks for compiling autograd",
+    -1,
+    _methods};
+
+static TraceState call_begin_capture(
+    PyObject* self,
+    CacheNode& cache,
+    AutogradCompilerCall& compiler_call,
+    size_t num_outputs) {
+  static PyObject* method_name = PyUnicode_InternFromString("begin_capture");
+  THPObjectPtr pyinput(THPVariable_WrapList(compiler_call.inputs));
+  THPObjectPtr pysizeinput(cache.wrap_dynamic_inputs());
+  THPObjectPtr pyresult(check(PyObject_CallMethodObjArgs(
+      self, method_name, pyinput.get(), pysizeinput.get(), NULL)));
+
+  PyObject *fake_inputs, *fake_sizes;
+  check(PyArg_ParseTuple(pyresult.get(), "OO", &fake_inputs, &fake_sizes));
+  return TraceState(
+      THPVariable_UnpackList(fake_inputs),
+      cache.unwrap_dynamic_inputs(fake_sizes),
+      compiler_call.accumulate_grad,
+      num_outputs);
+}
+
+static PyObject* call_end_capture(PyObject* self, const variable_list& inputs) {
+  static PyObject* method_name = PyUnicode_InternFromString("end_capture");
+  THPObjectPtr pyinput(THPVariable_WrapList(inputs));
+  return check(PyObject_CallMethodOneArg(self, method_name, pyinput.get()));
+}
+
+variable_list compiled_autograd(
+    const std::shared_ptr<Node>& graph_root,
+    GraphTask& graph_task,
+    bool accumulate_grad,
+    const edge_list& output_edges) {
+  TORCH_CHECK(
+      output_edges.empty() || !accumulate_grad,
+      "outputs+accumulate_grad not yet implemented")
+  static std::mutex lock;
+  std::lock_guard<std::mutex> lock_guard(lock);
+  pybind11::gil_scoped_acquire gil;
+  std::unordered_map<Node*, int> dependencies =
+      std::move(graph_task.dependencies_);
+  std::vector<std::shared_ptr<Node>> worklist{graph_root};
+  worklist.reserve(dependencies.size());
+  AutogradCompilerCall compiler_call(accumulate_grad);
+  for (const auto i : c10::irange(output_edges.size())) {
+    compiler_call.node_calls.lookup(output_edges[i].function)
+        .mark_output(output_edges[i].input_nr, i);
+  }
+  const bool check_exec_info = !graph_task.exec_info_.empty();
+  CacheNode* cache = CacheNode::root();
+  std::vector<NodeCall*> calls;
+  calls.reserve(dependencies.size() + 1);
```

Comment:
It's possible to reduce how much we reserve here now that we have exec_info

Reply:
Yes, but likely not worth it.  This line is a small perf optimization (avoid a bunch of calls to reallocate the vector memory), taking an extra pass to calculate the needed size might incur more cost that just dynamically resizing things.

Reply:
Oh not saying we should go through and check is_needed on all the nodes. Checking exec_info_.size() should be close to what we want.

Reply:
I'll change it to that.